### PR TITLE
fix: ensure fix-volumes uses node user on Mac

### PR DESCRIPTION
**Pull request by aldeed**
## Problem
Using `stat -c "%u:%g" .` to determine the UID:GID does not work when the host machine is Mac OSX and the working directory (`.`) is itself a volume linked to the host machine. On Mac, the command will always show root (0:0) as the owner because all volumes are always owned by root. This is true even if the underlying image has this directory pre-created with different ownership. The volume ownership masks the ownership built into the image.

## Solution
Since the true purpose of `fix-volumes.sh` is to ensure that all volumes are owned by the same user who will run the command, my fix is to pass this user name into the script as an argument. Currently we could assume "node" user, but the script is more versatile if it accepts the user name as an argument because only the calling Dockerfile/entrypoint truly knows which user it will be running the command as.

### Versioning
Because all images are rebuilt and retagged on every merge to `trunk`, all versions of `node-dev` images are currently affected by this bug. So in addition to making a `v4`, which makes it easy to force-pull this change into any project, I also updated all existing images to pass the new argument to `fix-volumes.sh`

## Testing
1. Run `./dockerfiles.sh build images/node-dev/12.10.0-v4/Dockerfile` in this repo
2. In API repo, latest `release-3.0.0` branch, change the image to `reactioncommerce/node-dev:12.10.0-v4` in `docker-compose.dev.yml`
3. If you haven't already, `ln -s docker-compose.dev.yml docker-compose.override.yml` in the API project.
4. In API project, `dc down -v`, or at least delete the node_modules volume with something like `docker volume rm reaction_api_node_modules`
5. In API project, `dc up -d mongo`
6. In API project, `dc up api`.

In the logs you should see something like this:

```
api_1    | fix-volumes: Fixing all volumes to be owned by node
api_1    | fix-volumes: Fixing volume /usr/local/src/app/node_modules (before=0:0 after=1000:1000)…✓
```

And NPM packages should install without errors and the API should start.

### CODE REVIEW

File: scripts/fix-volumes.sh [hunk 1]
```diff
@@ -13,26 +13,35 @@ set -o posix    # more strict failures in subshells
...
+
 volumes=$(mount | grep -E '^/dev/' | grep -Ev ' on /etc/' || true)
 if [[ -z "${volumes}" ]]; then
+  echo "fix-volumes: No volumes found"
   exit
 fi
-owner=$(stat -c "%u:%g" .)
+
+desired_owner_user=$1
+owner="$(id -u ${desired_owner_user}):$(id -g ${desired_owner_user})"
```
Comment (aldeed) [inline scripts/fix-volumes.sh]: This part is the main change in this PR, how we determine the owner

File: images/node-dev/12.10.0-v4/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
+FROM node:12.10.0-alpine
```
Comment (aldeed) [inline images/node-dev/12.10.0-v4/Dockerfile:1]: This file is identical to v3

File: images/node-dev/12.10.0-v4/scripts/entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env bash
```
Comment (aldeed) [inline images/node-dev/12.10.0-v4/scripts/entrypoint.sh:1]: This file is identical to v3

File: images/node-dev/10.16.3-v4/scripts/entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env bash
```
Comment (aldeed) [inline images/node-dev/10.16.3-v4/scripts/entrypoint.sh:1]: This file is identical to v3

File: images/node-dev/10.16.3-v4/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,22 @@
+FROM node:10.16.3-alpine
```
Comment (aldeed) [inline images/node-dev/10.16.3-v4/Dockerfile:1]: This file is identical to v3

File: scripts/fix-volumes.sh (hunk 1 above)
Comment (focusaurus) [reply scripts/fix-volumes.sh]: If I'm reading this diff right, we're now just hardcoding "node" which brings back the original mismatch issue on linux we had.

**Comment by focusaurus** [comment]
Is there any issue on mac? We used to just exit if we see uid 0 and do nothing, and is that not OK for mac?

**Comment by aldeed** [comment]
@focusaurus Yes the issue is that `stat -c "%u:%g" .` is always `0:0` so it exits. But we don't want it to exit. The app runs as `node` so the volumes all need to be changed to be owned by node.

**Comment by aldeed** [comment]
@focusaurus I'm not sure which exact change broke it on Mac, but the API was still back on `-v1` and updating to `-v3` it no longer works because of the issue I described.

File: scripts/fix-volumes.sh [hunk 6]
```diff
@@ -13,26 +13,47 @@ set -o posix    # more strict failures in subshells
 IFS=$'\n\t'
 # ---- End unofficial bash strict mode boilerplate
 
+if [[ "$#" -ne 1 ]]; then
+  echo "Usage: $0 [desired_owner_user]" 1>&2
+  exit 1
```
Comment (focusaurus) [inline scripts/fix-volumes.sh]: This is OK but I'd also think about can we make a default like:

`desired_owner_user="${1:-1000:1000}"` so it does something reasonably likely to work even without an argument?

File: scripts/fix-volumes.sh [hunk 7]
```diff
@@ -13,26 +13,47 @@ set -o posix    # more strict failures in subshells
...
+# be incorrectly carried in from the host machine. So the `if` below should be
+# true only on Mac.
+#
+# On Linux, we "fix" by copying the working directory ownership to all volumes,
+# in conjunction with changing the `node` user's UID to match the host machine,
+# which was done earlier, outside of this script.
+#
+# On Mac, we "fix" by setting directory ownership to the container `node` user's
+# UID:GID for every volume. The UID mismatch that happens sometimes on Linux
+# is never an issue on Mac.
```
Comment (focusaurus) [inline scripts/fix-volumes.sh:35]: Thanks for writing out this comment!

File: scripts/fix-volumes.sh [hunk 8]
```diff
@@ -13,26 +13,47 @@ set -o posix    # more strict failures in subshells
...
+  owner="$(id -u ${desired_owner_user}):$(id -g ${desired_owner_user})"
 fi
 
+echo "fix-volumes: Fixing all volumes to be owned by '${owner}'"
 echo "${volumes}" | awk '{print $3}' | {
   while IFS= read -r dir; do
-    echo "${dir}"
     mkdir -p "${dir}"
     old_owner=$(stat -c "%u:%g" "${dir}")
     if [[ "$1" != "--force" && "${old_owner}" == "${owner}" ]]; then
```
Comment (focusaurus) [inline scripts/fix-volumes.sh]: I think we need to fix up the positional argument handling here now that we're expecting an argument. `$1` is no longer the only argument.

**Review by focusaurus — CHANGES_REQUESTED** [review]

File: scripts/fix-volumes.sh (hunk 8 above)
Comment (aldeed) [reply scripts/fix-volumes.sh]: Didn't see that, thanks. Nothing uses --force so maybe I'll just remove this check for now?

**Comment by focusaurus** [comment]
Yeah I implemented `--force` initially because we only do a cheap check of
the top-level dir of each mount. You would need that in the case that the
mount point itself is OK but some subdirectory within it has the wrong
owner, so `--force` was "just make it work I don't care how long it takes".
It was intended for customer support scenario where regular fix-volumes.sh
was already running but not chowning anything because the wrongness was
only in subdirectories. I think it's useful still but don't want to block
the fix. We cloud trigger it with a command line flag with a bit more
parsing code or an env var or something eventually.

**Comment by aldeed** [comment]
I tested this on linux and I believe I was able to reproduce the issue and confirm the fix still works for linux, too. Here’s what I did:

```
sudo useradd testuser # create test user
sudo passwd testuser # set password for test user
usermod -a -G docker testuser # add test user to docker group so docker-compose command will work
chmod 644 .env # make .env visible to test user because we create it for current user only
(check uid/gid, 1001:1001 in my case)
sudo chown 1001:1001 . # change ownership of the api project root to new user/group
su - testuser # switch to new user
docker-compose up api # start api as test user
```

And then I saw this in the logs:

```
Fixing volume /usr/local/src/app/node_modules (before=1000:129 after=1001:1001)…✓ 
```

**Review by focusaurus — APPROVED** [review]

**Pull request merged**